Adding API route to set bgpPeerDescr field

geg347 opened the pull request:
DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Hello,

To add information about the BGP session in our alerts, we need to use the bgpPeerDescr field.
To ease the configuration of this field I created a new route in the API that update this value in database.

Maybe this feature will be usefull for others so I submitted it here.
Please just tell me if it needs some improvements or if you don't want this feature in the project.

Do not forget to clear the routes cached with "php artisan route:clear".

Bye

PipoCanaja commented on includes/html/api_functions.inc.php:
```diff
@@ -602,6 +602,39 @@ function get_bgp(Illuminate\Http\Request $request)
+    $bgp = [];
+    //find existing bgp for update
+    $bgpPeerId = $request->route('id');
+    $bgps = dbFetchRows("SELECT * FROM `bgpPeers` WHERE `bgpPeer_id` = $bgpPeerId LIMIT 1");
```
Hi @geg347 
Please have a look at the get_bgp function just above. A few checks are missing. You need to validate the data (if bgpPeerId is not numeric for instance) and most importantly the dbFetchRow must be injection_proof. Which is not the case here (using the question mark syntax).

geg347 commented:
Hello @PipoCanaja,
So I added the check on bgpPeerId to validate it is a integer.
I also changed the SQL request in dbFetchRow but I am not sure it is injection_proof. Do you have an example of an injection_proof dbFetchRow call, please?
And should I also check/validate the bgp_descr HTTP POST variable or the json_decode function is enough?

murrant commented:
@geg347 FYI, all db* function bypass any SQL injection protection. (because it is a compatibility shim)  Native Laravel queries are much safer and more portable.

geg347 commented:
Hello @murrant,
OK so the dbFetchRows function should be safe by testing if the id is numeric.
But for the dbUpdate I copy paste code from existing API function for editing bills.
I don't know how to protect this from SQL injection... Should I edit the common function dbFacile to prepare the SQL request as PDO can do? Is there some other tests possible in the edit_bgp_descr function?

geg347 commented:
Hello,
Any update on this please?
What can I do, do you have a working example?

murrant commented on includes/html/api_functions.inc.php:
```diff
@@ -602,6 +602,43 @@ function get_bgp(Illuminate\Http\Request $request)
+        return api_error(400, 'Invalid id has been provided');
+    }
+
+    $bgps = dbFetchRows('SELECT * FROM `bgpPeers` WHERE `bgpPeer_id` = ?', [$bgpPeerId]);
```
$peer = \App\Models\BgpPeer::firstWhere('bgpPeer_id', $bgpPeerId);

murrant commented on includes/html/api_functions.inc.php:
```diff
@@ -602,6 +602,43 @@ function get_bgp(Illuminate\Http\Request $request)
+    $bgp_descr = $data['bgp_descr'];
+
+    // update existing bgp
+    if (count($bgps) == 0) {
```
if ($peer === null) {

murrant commented on includes/html/api_functions.inc.php:
```diff
@@ -602,6 +602,43 @@ function get_bgp(Illuminate\Http\Request $request)
+    // update existing bgp
+    if (count($bgps) == 0) {
+        return api_error(404, 'BGP peer ' . $bgpPeerId . ' does not exist');
+    } elseif (is_array($bgps) && count($bgps) == 1) {
```
delete this elseif we already know one was found

murrant commented on includes/html/api_functions.inc.php:
```diff
@@ -602,6 +602,43 @@ function get_bgp(Illuminate\Http\Request $request)
+    } elseif (is_array($bgps) && count($bgps) == 1) {
+        $bgp = $bgps[0];
+
+        $update_data = [
```
$peer->bgpPeerDescr = $bgp_descr;

murrant commented on includes/html/api_functions.inc.php:
```diff
@@ -602,6 +602,43 @@ function get_bgp(Illuminate\Http\Request $request)
+        $update_data = [
+            'bgpPeerDescr' => $bgp_descr,
+        ];
+        $update = dbUpdate($update_data, 'bgpPeers', 'bgpPeer_id=?', [$bgpPeerId]);
```
if ($peer->save()) {

geg347 replied:
I have put a last line : return api_error(500, 'Failed to update existing bgp');
I don't know if it is usefull or if it can cause some problem.

murrant replied:
Yeah, better to have more error messages than less to help the api user.

PipoCanaja commented:
All good then @geg347
Thx for the contribution.
